patches/tModLoader/Terraria/Recipe.cs.patch

```diff
@@ -473,8 +488,29 @@
 +			if (Main.recipe[i].Disabled)
 +				continue;
 +
- 			if (!Main.recipe[i].notDecraftable)
- 				ItemID.Sets.IsCrafted[Main.recipe[i].createItem.type] = i;
++			/*Vanilla:
+-			if (!Main.recipe[i].notDecraftable)
++			 * if (!Main.recipe[i].notDecraftable)
+-				ItemID.Sets.IsCrafted[Main.recipe[i].createItem.type] = i;
++			 * ItemID.Sets.IsCrafted[Main.recipe[i].createItem.type] = i;
++			 */
+ 
++			if (!Main.recipe[i].notDecraftable) {
```

Comment:
Can you use a `List<int>[]` internally here, and then `ToArray` to create `ItemID.Sets.IsCrafted`? Better for performance and readability. Can use `null` lists for items which have no recipe initially. 

`(recipeLists[itemType] ??= new()).Add(i);`

Reply:
Aight so I might have what you're asking wrong, but is:

```
		List<int>[] recipeLists = new List<int>[ItemLoader.ItemCount];
		for (int i = 0; i < numRecipes; i++) {

			if (Main.recipe[i].Disabled)
				continue;

			if (!Main.recipe[i].notDecraftable)
				(recipeLists[Main.recipe[i].createItem.type] ??= new()).Add(i);

		}
		
		ItemID.Sets.CraftingRecipeIndeces = recipeLists.Select(l => l.ToArray()).ToArray();
```

What you mean? and if so do I remove the defualt state from the factory? since it should no longer be needed

Reply:
Looks good to me! Yes, remove the default from the factory, let it be null. Shouldn't be accessed prior to this calculation.

Also, typo in your snippet above. `Indices` rather than `Indeces`

Reply:
Also, please do `var r = Main.recipe[i]`. Repeated array access triggers me a little :')

Reply:
👍